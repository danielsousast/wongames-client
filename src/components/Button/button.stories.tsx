import { Meta, StoryObj } from '@storybook/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'
import Button from '.'
import React from 'react'

export default {
  title: 'Button',
  component: Button
} as Meta

export const Default: StoryObj = {
  args: {
    children: 'Button',
    size: 'medium'
  }
}

export const OnlyIcon: StoryObj = {
  args: {
    children: null,
    size: 'small',
    icon: <AddShoppingCart />
  }
}

export const IconAndText: StoryObj = {
  args: {
    children: 'Button',
    size: 'small',
    icon: <AddShoppingCart />
  }
}

export const AsLink: StoryObj = {
  args: {
    children: 'Button',
    size: 'large',
    as: 'a',
    href: '/link'
  }
}
