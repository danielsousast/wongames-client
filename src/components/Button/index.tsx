import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'
import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  $fullWidth?: boolean
  minimal?: boolean
  icon?: JSX.Element
  as?: React.ElementType
} & ButtonTypes

const Button = ({
  children,
  size = 'medium',
  $fullWidth = false,
  icon,
  minimal = false,
  ...props
}: ButtonProps) => {
  return (
    <S.ButtonWrapper
      size={size}
      $fullWidth={$fullWidth}
      $hasIcon={!!icon}
      $minimal={minimal}
      {...props}
    >
      {!!icon && icon}
      {!!children && <span>{children}</span>}
    </S.ButtonWrapper>
  )
}

export default Button
