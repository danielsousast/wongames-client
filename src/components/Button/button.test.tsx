import { render, screen } from '@testing-library/react'
import 'jest-styled-components'
import Button from '.'
import { renderWithTheme } from '../../utils/tests/helpers'
import { AddShoppingCart } from 'styled-icons/material-outlined'

describe('<Button />', () => {
  it('should render the medium sizew by defaut', () => {
    const { container } = renderWithTheme(<Button>Button</Button>)
    expect(screen.getByRole('button', { name: /button/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem'
    })
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Button</Button>)
    expect(screen.getByRole('button', { name: /button/i })).toHaveStyle({
      height: '3rem',
      padding: '0.8rem',
      'font-size': '1.2rem'
    })
  })

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Button</Button>)
    expect(screen.getByRole('button', { name: /button/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.6rem'
    })
  })

  it('should render a fullWidth version', () => {
    renderWithTheme(<Button $fullWidth>Button</Button>)
    expect(screen.getByRole('button', { name: /button/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Button</Button>
    )
    expect(screen.getByText(/button/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render an only icon version', () => {
    renderWithTheme(<Button icon={<AddShoppingCart data-testid="icon" />} />)
    expect(screen.getByTestId('icon')).toBeInTheDocument()
    expect(screen.queryByText(/button/i)).not.toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Buy now
      </Button>
    )

    expect(screen.getByRole('link', { name: /buy now/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })

  it('should render a minimal version', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />} minimal>
        Button
      </Button>
    )
    expect(screen.getByRole('button', { name: /button/i })).toHaveStyle({
      background: 'none',
      color: '#F231A5'
    })
    expect(screen.getByRole('button', { name: /button/i })).toHaveStyleRule(
      'background',
      'none',
      { modifier: ':hover' }
    )
  })
})
