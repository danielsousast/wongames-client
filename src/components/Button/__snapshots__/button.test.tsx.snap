// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium sizew by defaut 1`] = `
.c0 {
  border: 0;
  cursor: pointer;
  color: #FAFAFA;
  padding: 0.8rem;
  border-radius: 0.4rem;
  text-decoration: none;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
}

<button
  class="c0"
>
  <span>
    Button
  </span>
</button>
`;
